src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java

```diff
@@ -916,10 +933,61 @@ private static void applyRestrictionsToNewEdge(
                 "Recreate new restriction {} with split edge as to edge {}", splitTurnRestriction,
                 newEdge
         );
-        graph.addTurnRestriction(restriction.from, splitTurnRestriction);
+        restriction.from.addTurnRestriction(splitTurnRestriction);
         // Former turn restriction needs to be removed. Especially no only_turn
-        // restriction to a non existent edge must not survive
-        graph.removeTurnRestriction(restriction.from, restriction);
+        // restriction to a non-existent edge must not survive
+        restriction.from.removeTurnRestriction(restriction);
+    }
+
+    /**
+     * Add a {@link TurnRestriction} to this edge.
+     *
+     * This method is thread-safe as modifying the underlying set is synchronized.
+     */
+    public void addTurnRestriction(TurnRestriction turnRestriction) {
+        if (turnRestriction == null) { return; }
+        synchronized (this) {
+            // in order to guarantee fast access without extra allocations
+            // we make the turn restrictions unmodifiable after a copy-on-write modification
+            var temp = new HashSet<>(turnRestrictions);
+            temp.add(turnRestriction);
+            turnRestrictions = Collections.unmodifiableSet(temp);
```

Comment:
```suggestion
            var temp = new ArrayList<>(turnRestrictions);
            temp.add(turnRestriction);
            turnRestrictions = Set.copyOf(temp);
```
If the set do not contain `null` then you should use `copyOf` witch create an unmodifiable collection and not a view around the HashSet. This will be much more efficient if the sets are small.